server/src/main/java/org/elasticsearch/search/retriever/RankDocsRetrieverBuilder.java

```diff
@@ -79,43 +79,49 @@ public RetrieverBuilder rewrite(QueryRewriteContext ctx) throws IOException {
+        BoolQueryBuilder boolQuery = new BoolQueryBuilder();
         for (var retriever : sources) {
             var query = retriever.topDocsQuery();
             if (query != null) {
                 if (retriever.retrieverName() != null) {
                     query.queryName(retriever.retrieverName());
                 }
-                disMax.add(query);
+                boolQuery.should(query);
             }
         }
         // ignore prefilters of this level, they are already propagated to children
-        return disMax;
+        return boolQuery;
     }
 
     @Override
     public void extractToSearchSourceBuilder(SearchSourceBuilder searchSourceBuilder, boolean compoundUsed) {
         // here we force a custom sort based on the rank of the documents
-        // TODO: should we adjust to account for other fields sort options just for the top ranked docs?
         if (searchSourceBuilder.rescores() == null || searchSourceBuilder.rescores().isEmpty()) {
-            searchSourceBuilder.sort(Arrays.asList(new RankDocsSortBuilder(rankDocs.get()), new ScoreSortBuilder()));
-        }
-        if (searchSourceBuilder.explain() != null && searchSourceBuilder.explain()) {
-            searchSourceBuilder.trackScores(true);
+            searchSourceBuilder.sort(List.of(new RankDocsAndScoreSortBuilder(rankDocs.get())));
         }
+
         BoolQueryBuilder boolQuery = new BoolQueryBuilder();
         RankDocsQueryBuilder rankQuery = new RankDocsQueryBuilder(rankDocs.get());
         // if we have aggregations we need to compute them based on all doc matches, not just the top hits
-        // so we just post-filter the top hits based on the rank queries we have
-        if (searchSourceBuilder.aggregations() != null) {
+        // similarly, for profile and explain we re-run all parent queries to get all needed information
+        if (hasAggregations(searchSourceBuilder)
+            || isExplainRequest(searchSourceBuilder)
+            || isProfileRequest(searchSourceBuilder)
+            || shouldTrackTotalHits(searchSourceBuilder)) {
             boolQuery.should(rankQuery);
             // compute a disjunction of all the query sources that were executed to compute the top rank docs
             QueryBuilder disjunctionOfSources = topDocsQuery();
             if (disjunctionOfSources != null) {
-                boolQuery.should(disjunctionOfSources);
+                // if have an explain or profile request, we'd need to compute scores for all documents and retrieve detailed info for
+                // query execution times
+                if (isExplainRequest(searchSourceBuilder) || isProfileRequest(searchSourceBuilder)) {
+                    boolQuery.must(disjunctionOfSources);
+                    searchSourceBuilder.trackScores(true);
+                } else {
+                    // otherwise we apply all the parent queries as filters
+                    boolQuery.filter(disjunctionOfSources);
```

Comment:
The score could be required by aggregations too. I think we can always use a `must` here, if the score is not required the searcher will do the optimisation locally.

Reply:
> The score could be required by aggregations too.

Ah fair enough! I have a few tests with aggs that seem to work as expected. What would the scenario be in this case so that I can capture it in a test?

Reply:
Scripts are the first example that comes to mind:
```
"aggs": {
        "max_score": {
          "max": {
            "script": {
              "lang": "painless",
              "inline": "_score"
            }
          }
        }
```